lib/tezos_client/smartpy_interface.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+require_relative "smartpy_inteface/smartpy_wrapper"
+require_relative "smartpy_inteface/entry_point_serializer_wrapper"
+require_relative "smartpy_inteface/micheline_serializer_wrapper"
+
+class TezosClient
+  class SmartpyInterface
+    include Logger
+    include SmartpyWrapper
+    include MichelineSerializerWrapper
+    include EntryPointSerializerWrapper
+
+    attr_reader :options
+
+    def initialize(rpc_node_address: "127.0.0.1", rpc_node_port: 8732, options: {})
+      @rpc_node_address = rpc_node_address
+      @rpc_node_port = rpc_node_port
+      @options = options
+    end
+
+    def json_scripts(args)
+      compile_to_michelson(args) do |contract_script_filename, init_script_filename|
+        micheline_contract = convert_michelson_to_micheline(read_file(contract_script_filename))
+        micheline_storage = convert_michelson_to_micheline(read_file(init_script_filename))
+
+        [JSON.parse(micheline_storage), JSON.parse(micheline_contract)]
+      end
+    end
+
+    def origination_script(args)
+      json_init_script, json_contract_script = json_scripts(args)
+
+      {
+        code: json_contract_script,
+        storage: json_init_script
+      }
+    end
+
+    def call_parameters(args)
+      compile_to_michelson(args) do |contract_script_filename, _|
+        params_struct = read_file(contract_script_filename).split("\n").first
+        entry_point = args[:parameters].first
+        params = args[:parameters][1..-1]
+
+        JSON.parse(gen_entry_point_args(params_struct, entry_point, params))
+      end
+    end
+
+    private
+
+    def compile_to_michelson(args)
+      with_file_copy(args[:script]) do |script_copy_path|
+        script_basename = script_copy_path.split("/").last.sub(/.py$/, "")
+        script_path = "/tmp/#{script_basename}/"
+        init_script_filename = "contractStorage.tz"
+        contract_script_filename = "contractCode.tz"
+        call_smartpy ["local-compile", script_copy_path, args[:init_params], script_path]
+
+        yield(script_path + contract_script_filename, script_path + init_script_filename)
+      end
+    end
+
+    def with_tempfile(extension)
```

Comment:
with_tempfile, with_file_copy and read_file could be set in a module in order to factorize code.